File: sdk/nodejs/policy/deserialize.ts
// Copyright 2016-2019, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { asset, Output } from "@pulumi/pulumi";
import {
    specialArchiveSig,
    specialAssetSig,
    specialSecretSig,
    specialSigKey,
} from "@pulumi/pulumi/runtime/rpc";

import { Secret } from "./policy";
import { isSpecialProxy, getSpecialProxyTarget } from "./proxy";
import { secretsPreservingProxy } from "./secret";

/**
 * deserializeProperties fetches the raw outputs and deserializes them from a gRPC call result.
 * @internal
 */
export function deserializeProperties(outputsStruct: any, proxySecrets: boolean): any {
    let props: any = {};
    const outputs: any = outputsStruct.toJavaScript();
    for (const k of Object.keys(outputs)) {
        // We treat properties with undefined values as if they do not exist.
        if (outputs[k] !== undefined) {
            props[k] = deserializeProperty(outputs[k], proxySecrets);
        }
    }
    if (proxySecrets) {
        props = secretsPreservingProxy(props);
    }
    return props;
}

/**
 * deserializeProperty is partly based on, but not functionally or superficially identical to, the
 * deserialization code in `@pulumi/pulumi/runtime/rpc.ts`.
 *
 * Specifically, we explicitly do not want to be compatible with Output<T>. Policies interact with
 * raw JSON representations of resource definitions, so things like secrets must be "fully unpacked"
 * into plain-old-JSON that the policies can interact with as normal data. If things like secrets
 * are not unpacked, every field in every resource type would have to be type `T | Secret<T>`, which
 * would severely detract from usability.
 */
function deserializeProperty(prop: any, proxySecrets: boolean): any {
    if (prop === undefined) {
        throw new Error("unexpected undefined property value during deserialization");
    } else if (
        prop === null ||
        typeof prop === "boolean" ||
        typeof prop === "number" ||
        typeof prop === "string"
    ) {
        return prop;
    } else if (prop instanceof Array) {
        const elems: any[] = [];
        for (const e of prop) {
            elems.push(deserializeProperty(e, proxySecrets));
        }

        return elems;
    } else {
        // We need to recognize assets and archives specially, so we can produce the right runtime
        // objects.
        const sig: any = prop[specialSigKey];
        if (sig) {
            switch (sig) {
                case specialAssetSig:
                    if (prop["path"]) {
                        return new asset.FileAsset(<string>prop["path"]);
                    } else if (prop["text"]) {
                        return new asset.StringAsset(<string>prop["text"]);
                    } else if (prop["uri"]) {
                        return new asset.RemoteAsset(<string>prop["uri"]);
                    } else {
                        throw new Error(
                            "Invalid asset encountered when unmarshaling resource property",
                        );
                    }
                case specialArchiveSig:
                    if (prop["assets"]) {
                        const assets: asset.AssetMap = {};
                        for (const name of Object.keys(prop["assets"])) {
                            const a = deserializeProperty(prop["assets"][name], proxySecrets);
                            if (!asset.Asset.isInstance(a) && !asset.Archive.isInstance(a)) {
                                throw new Error(
                                    "Expected an AssetArchive's assets to be unmarshaled Asset or Archive objects",
                                );
                            }
                            assets[name] = a;
                        }
                        return new asset.AssetArchive(assets);
                    } else if (prop["path"]) {
                        return new asset.FileArchive(<string>prop["path"]);
                    } else if (prop["uri"]) {
                        return new asset.RemoteArchive(<string>prop["uri"]);
                    } else {
                        throw new Error(
                            "Invalid archive encountered when unmarshaling resource property",
                        );
                    }
                case specialSecretSig:
                    let value = deserializeProperty(prop["value"], proxySecrets);
                    if (proxySecrets) {
                        // Wrap the value so that a proxy wrapper can detect it later on.
                        value = new Secret(value);
                    }
                    return value;
                default:
                    throw new Error(
                        `Unrecognized signature '${sig}' when unmarshaling resource property`,
                    );
            }
        }

        const obj: any = {};
        for (const k of Object.keys(prop)) {
            obj[k] = deserializeProperty(prop[k], proxySecrets);
        }

        return obj;
    }
}

/**
 * serializeProperties serializes a runtime resource object to ready it for a gRPC call result.
 * @internal
 */
export async function serializeProperties(obj: any): Promise<any> {
    return serializeProperty(obj);
}

/**
 * serializeProperty is partly based on, but not functionally or superficially identical to, the
 * serialization code in `@pulumi/pulumi/runtime/rpc.ts`. It has to handle the slightly different
 * serialization semantics for policies which treat outputs and secrets with different semantics.
 */
async function serializeProperty(prop: any): Promise<any> {
    if (prop === undefined) {
        return null;
    }
    if (prop === null ||
            typeof prop === "boolean" ||
            typeof prop === "number" ||
            typeof prop === "string") {
        return prop;
    }
    if (prop[isSpecialProxy]) {
        return await serializeProperty(prop[getSpecialProxyTarget]);
    }
    if (prop instanceof Promise) {
        return serializeProperty(await prop);
    }
    if (prop instanceof Array) {
        const elems: any[] = [];
        for (const e of prop) {
            const se = await serializeProperty(e);
            elems.push(se === undefined ? null : se);
        }
        return elems;
    }
    if (prop instanceof asset.FileAsset) {
        return { [specialSigKey]: specialAssetSig, path: await serializeProperty(prop.path) };
    }
    if (prop instanceof asset.StringAsset) {
        return { [specialSigKey]: specialAssetSig, text: await serializeProperty(prop.text) };
    }
    if (prop instanceof asset.RemoteAsset) {
        return { [specialSigKey]: specialAssetSig, uri: await serializeProperty(prop.uri) };
    }
    if (prop instanceof asset.AssetArchive) {
        const assets: any[] = [];
        for (const a of Object.keys(await prop.assets)) {
            assets.push(await serializeProperty(a));
        }
        return { [specialSigKey]: specialArchiveSig, assets };
    }
    if (prop instanceof asset.FileArchive) {
        return { [specialSigKey]: specialArchiveSig, path: await serializeProperty(prop.path) };
    }
    if (prop instanceof asset.RemoteArchive) {
        return { [specialSigKey]: specialArchiveSig, uri: await serializeProperty(prop.uri) };
    }
    if (prop instanceof Secret) {
        // Because of the way secrets proxying works, we very well may encounter a
        // secret in its raw form, since serialization explicitly unwraps the proxy and
        // accesses the raw underlying values.
        return {
            [specialSigKey]: specialSecretSig,
            value: await serializeProperty(prop.value),
        };
    }

    // Unsupported types:
    if (Output.isInstance(prop)) {
        throw new Error("Serializing output values not supported from within a policy pack");
    }

    const obj: any = {};
    for (const k of Object.keys(prop)) {
        const value = await serializeProperty(prop[k]);
        if (value !== undefined) {
            obj[k] = value;
        }
    }
    return obj;
}

File: sdk/python/lib/pulumi_policy/deserialize.py
# Copyright 2016-2020, Pulumi Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Any, Dict, List, Mapping, Union

import pulumi
from pulumi.runtime.rpc import isLegalProtobufValue

from .proxy import _DictProxy, _ListProxy
from .secret import _DictSecretsProxy, _ListSecretsProxy, Secret, secrets_preserving_proxy

# SPECIAL_SIG_KEY is sometimes used to encode type identity inside of a map.
# See https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/properties.go.
SPECIAL_SIG_KEY = "4dabf18193072939515e22adb298388d"

# SPECIAL_ASSET_SIG is a randomly assigned hash used to identify assets in maps.
# See https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/properties.go.
SPECIAL_ASSET_SIG = "c44067f5952c0a294b673a41bacd8c17"

# SPECIAL_ARCHIVE_SIG is a randomly assigned hash used to identify archives in maps.
# See https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/properties.go.
SPECIAL_ARCHIVE_SIG = "0def7320c3a5731c473e5ecbe6d01bc7"

# SPECIAL_SECRET_SIG is a randomly assigned hash used to identify secrets in maps.
# See https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/properties.go.
SPECIAL_SECRET_SIG = "1b47061264138c4ac30d75fd1eb44270"

def deserialize_properties(props: Dict[str, Any], proxy_secrets: bool = False) -> Dict[str, Any]:
    """
    Deserializes properties from a gRPC call result.
    """
    result: Dict[str, Any] = {}
    for key in props:
        result[key] = _deserialize_property(props[key], proxy_secrets)

    if proxy_secrets:
        result = secrets_preserving_proxy(result)

    return result

def _deserialize_property(prop: Any, proxy_secrets: bool = False) -> Any:
    if isinstance(prop, list):
        elems: List[Any] = []
        for e in prop:
            elems.append(_deserialize_property(e, proxy_secrets))
        return elems
    if isinstance(prop, dict):
        if SPECIAL_SIG_KEY in prop:
            sig = prop[SPECIAL_SIG_KEY]
            if sig == SPECIAL_ASSET_SIG:
                return _deserialize_asset(prop)
            if sig == SPECIAL_ARCHIVE_SIG:
                return _deserialize_archive(prop, proxy_secrets)
            if sig == SPECIAL_SECRET_SIG:
                if "value" not in prop:
                    raise AssertionError("Invalid secret encountered when unmarshaling resource property")
                value = _deserialize_property(prop["value"], proxy_secrets)
                if proxy_secrets:
                    # Mark the secret so that a proxy wrapper can detect it later on.
                    value = Secret(value)
                return value
            raise AssertionError(f"Unrecognized signature '{sig}' when unmarshaling resource property")
        return deserialize_properties(prop, proxy_secrets)
    return prop

def _deserialize_asset(prop: Dict[str, Any]) -> pulumi.Asset:
    if "path" in prop:
        return pulumi.FileAsset(prop["path"])
    if "text" in prop:
        return pulumi.StringAsset(prop["text"])
    if "uri" in prop:
        return pulumi.RemoteAsset(prop["uri"])
    raise AssertionError("Invalid asset encountered when unmarshaling resource property")

def _deserialize_archive(prop: Dict[str, Any], proxy_secrets: bool) -> pulumi.Archive:
    if "assets" in prop:
        assets: Dict[str, Union[pulumi.Asset, pulumi.Archive]] = {}
        for key in prop["assets"]:
            a = _deserialize_property(prop["assets"][key], proxy_secrets)
            if not isinstance(a, pulumi.Asset) and not isinstance(a, pulumi.Archive):
                raise AssertionError("Expected an AssetArchive's assets to be unmarshaled Asset or Archive objects")
            assets[key] = a
        return pulumi.AssetArchive(assets)
    if "path" in prop:
        return pulumi.FileArchive(prop["path"])
    if "uri" in prop:
        return pulumi.RemoteArchive(prop["uri"])
    raise AssertionError("Invalid archive encountered when unmarshaling resource property")

def serialize_properties(props: Mapping[str, Any]) -> Mapping[str, Any]:
    """
    Serializes properties in preparation for a gRPC call.
    """
    return _serialize_property(props)

def _serialize_property(prop: Any) -> Any:
    # Check for proxies, and unwrap to get the raw elements. This ensures unknowns flow through.
    if isinstance(prop, _DictProxy):
        return _serialize_property(prop["__target"])
    if isinstance(prop, _ListProxy):
        return _serialize_property(prop["__target"])
    if isinstance(prop, _DictSecretsProxy):
        return _serialize_property(prop["__target"])
    if isinstance(prop, _ListSecretsProxy):
        return _serialize_property(prop["__target"])

    if isinstance(prop, list):
        elems: List[Any] = []
        for e in prop:
            elems.append(_serialize_property(e))
        return elems

    # Check for assets:
    if isinstance(prop, pulumi.FileAsset):
        return { SPECIAL_SIG_KEY: SPECIAL_ASSET_SIG, "path": _serialize_property(prop.path) }
    if isinstance(prop, pulumi.StringAsset):
        return { SPECIAL_SIG_KEY: SPECIAL_ASSET_SIG, "text": _serialize_property(prop.text) }
    if isinstance(prop, pulumi.RemoteAsset):
        return { SPECIAL_SIG_KEY: SPECIAL_ASSET_SIG, "uri": _serialize_property(prop.uri) }

    # Check for archives:
    if isinstance(prop, pulumi.AssetArchive):
        assets: Dict[str, Dict[str, Any]] = {}
        for key in prop.assets:
            assets[key] = _serialize_property(prop.assets[key])
        return { SPECIAL_SIG_KEY: SPECIAL_ARCHIVE_SIG, "assets": assets }
    if isinstance(prop, pulumi.FileArchive):
        return { SPECIAL_SIG_KEY: SPECIAL_ARCHIVE_SIG, "path": _serialize_property(prop.path) }
    if isinstance(prop, pulumi.RemoteArchive):
        return { SPECIAL_SIG_KEY: SPECIAL_ARCHIVE_SIG, "uri": _serialize_property(prop.uri) }

    # Check for secrets:
    if isinstance(prop, Secret):
        # Because of the way secrets proxying works, we very well may encounter a
        # secret in its raw form, since serialization explicitly unwraps the proxy and
        # accesses the raw underlying value.
        return { SPECIAL_SIG_KEY: SPECIAL_SECRET_SIG, "value": _serialize_property(prop.value) }

    if isinstance(prop, Mapping):
        result: Dict[str, Any] = {}
        for key in prop:
            result[key] = _serialize_property(prop[key])
        return result

    if not isLegalProtobufValue(prop):
        raise TypeError(f"Serializing values of type {type(prop).__name__} not supported from within a policy pack")

    return prop
